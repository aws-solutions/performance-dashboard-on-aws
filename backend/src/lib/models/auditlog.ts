/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
