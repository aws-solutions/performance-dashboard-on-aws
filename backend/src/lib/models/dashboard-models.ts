export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};