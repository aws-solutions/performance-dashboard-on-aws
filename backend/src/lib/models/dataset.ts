/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export interface Dataset {
    id: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: Date;
    sourceType: SourceType;
    schema?: DatasetSchema;
}

export type DatasetContent = Array<object>;
export type DatasetList = Array<Dataset>;

export interface DatasetItem {
    pk: string;
    sk: string;
    type: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: string;
    sourceType: string;
    schema?: string;
}
