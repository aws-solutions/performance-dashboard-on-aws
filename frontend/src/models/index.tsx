/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
