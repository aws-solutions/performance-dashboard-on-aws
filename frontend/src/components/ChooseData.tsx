/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useState, MouseEvent } from "react";
import { useDateTimeFormatter, useWindowSize } from "../hooks";
import { Dataset, DatasetType } from "../models";
import Button from "./Button";
import RadioButtonsTile from "./RadioButtonsTile";
import FileInput from "./FileInput";
import Link from "./Link";
import Table from "./Table";
import { useTranslation } from "react-i18next";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch } from "@fortawesome/free-solid-svg-icons";
import styles from "./ChooseData.module.scss";

interface Props {
    handleChange: React.FormEventHandler<HTMLFieldSetElement>;
    datasetType: DatasetType | undefined;
    register: Function;
    onCancel: (event: MouseEvent<HTMLButtonElement>) => void;
    backStep: (event: MouseEvent<HTMLButtonElement>) => void;
    advanceStep: Function;
    fileLoading: boolean;
    csvErrors: Array<object> | undefined;
    csvFile: File | undefined;
    staticFileName: string | undefined;
    dynamicFileName: string | undefined;
    onFileProcessed: Function;
    browseDatasets: (event: MouseEvent<HTMLButtonElement>) => void;
    selectDynamicDataset: Function;
    dynamicDatasets: Array<Dataset>;
    hasErrors: boolean;
    widgetType: string;
    setShowNoDatasetTypeAlert: Function;
}

function ChooseData(props: Props) {
    const { t } = useTranslation();
    const dateFormatter = useDateTimeFormatter();
    const [filter, setFilter] = useState("");
    const [query, setQuery] = useState("");
    const windowSize = useWindowSize();
    const isMobile = windowSize.width <= 600;

    const onSelect = useCallback(
        (selectedDataset: Array<Dataset>) => {
            if (props.datasetType === DatasetType.DynamicDataset) {
                if (selectedDataset.length) {
                    props.selectDynamicDataset(selectedDataset[0]);
                    props.setShowNoDatasetTypeAlert(false);
                } else if (props.dynamicFileName) {
                    props.setShowNoDatasetTypeAlert(false);
                }
            }
            if (
                props.datasetType === DatasetType.StaticDataset &&
                (props.csvFile || props.staticFileName)
            ) {
                props.setShowNoDatasetTypeAlert(false);
            }
        },
        [props.datasetType],
    );

    const onSearch = (query: string) => {
        setFilter(query);
    };

    const onAdvanceStep = () => {
        if (props.hasErrors) {
            props.setShowNoDatasetTypeAlert(true);
        } else {
            props.advanceStep();
        }
    };

    return (
        <>
            <fieldset className="usa-fieldset" onChange={props.handleChange}>
                <legend className={`usa-hint ${isMobile ? "grid-col-12" : "grid-col-6"}`}>
                    <label className="usa-label text-bold">{t("Data")}</label>
                    <span className="usa-hint">
                        {t("ChooseDataDescription", {
                            widgetType: props.widgetType,
                        })}{" "}
                        <Link to="/admin/apihelp" target="_blank" external>
                            {t("HowDoIAddDatasets")}
                        </Link>
                    </span>
                </legend>

                <div className="grid-row">
                    <RadioButtonsTile
                        isHorizontally={true}
                        register={props.register}
                        options={[
                            {
                                id: "staticDataset",
                                value: "StaticDataset",
                                name: "datasetType",
                                dataTestId: "staticDatasetRadioButton",
                                label: t("StaticDataset"),
                                description: t("StaticDatasetDescription"),
                            },
                            {
                                id: "dynamicDataset",
                                value: "DynamicDataset",
                                name: "datasetType",
                                dataTestId: "dynamicDatasetRadioButton",
                                label: t("DynamicDataset"),
                                description: t("DynamicDatasetDescription"),
                            },
                        ]}
                    />
                </div>

                <div
                    hidden={props.datasetType !== DatasetType.StaticDataset}
                    role="tabpanel"
                    tabIndex={0}
                    aria-label={t("ChooseStaticDataset")}
                >
                    <div className="grid-row">
                        <div className="tablet:grid-col-6">
                            <FileInput
                                id="dataset"
                                name="dataset"
                                label={t("StaticDatasets")}
                                accept=".csv,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
                                loading={props.fileLoading}
                                errors={props.csvErrors}
                                register={props.register}
                                hint={
                                    <span>
                                        {t("StaticDatasetsHint")}{" "}
                                        <Link to="/admin/formatting" target="_blank" external>
                                            {t("HowDoIFormatMyCSVFile")}
                                        </Link>
                                    </span>
                                }
                                fileName={props.csvFile && props.csvFile.name}
                                staticFileName={props.staticFileName}
                                onFileProcessed={props.onFileProcessed}
                            />
                        </div>
                        <div className="tablet:grid-col-3">
                            <Button
                                variant="outline"
                                type="button"
                                className={isMobile ? "margin-top-1" : styles.vertCenter}
                                onClick={props.browseDatasets}
                            >
                                {t("BrowseDatasets")}
                            </Button>
                        </div>
                    </div>
                </div>

                <div
                    hidden={props.datasetType !== DatasetType.DynamicDataset}
                    role="tabpanel"
                    tabIndex={0}
                    aria-label={t("ChooseDynamicDataset")}
                >
                    <div className="grid-row margin-top-3 margin-bottom-1">
                        <div className="tablet:grid-col-4 padding-top-1px">
                            <label htmlFor="search">{`${t("Search.SearchDynamicDatasets")}`}</label>
                            <div role="search" className="usa-search usa-search--small">
                                <input
                                    className="usa-input"
                                    id="search"
                                    type="search"
                                    name="query"
                                    aria-label={t("ARIA.SearchInput")}
                                    style={{ height: "37px" }}
                                    value={query}
                                    placeholder={props.dynamicFileName}
                                    onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
                                        setQuery(event.target.value)
                                    }
                                />
                                <button
                                    className="usa-button usa-button--base padding-x-2"
                                    type="button"
                                    aria-label={t("ARIA.SearchButton")}
                                    style={{
                                        height: "37px",
                                        borderTopLeftRadius: "0",
                                        borderBottomLeftRadius: "0",
                                    }}
                                    onClick={() => onSearch(query)}
                                >
                                    <FontAwesomeIcon
                                        style={{ marginTop: "-3px" }}
                                        icon={faSearch}
                                    />
                                </button>
                            </div>
                        </div>
                    </div>

                    <div className="overflow-hidden">
                        <Table
                            id="items"
                            selection="single"
                            initialSortByField="updatedAt"
                            filterQuery={filter}
                            rows={React.useMemo(
                                () => props.dynamicDatasets,
                                [props.dynamicDatasets],
                            )}
                            screenReaderField="name"
                            width="100%"
                            pageSize={50}
                            onSelection={onSelect}
                            columns={React.useMemo(
                                () => [
                                    {
                                        Header: t("NameUpperCase"),
                                        accessor: "fileName",
                                    },
                                    {
                                        Header: t("LastUpdatedLabel"),
                                        accessor: "updatedAt",
                                        Cell: (props: any) => dateFormatter(props.value),
                                    },
                                    {
                                        Header: t("Description"),
                                        accessor: "description",
                                    },
                                    {
                                        Header: t("Tags"),
                                        accessor: "tags",
                                    },
                                ],
                                [dateFormatter, t],
                            )}
                        />
                    </div>
                </div>
            </fieldset>
            <br />
            <br />
            <hr />
            <Button
                variant="outline"
                type="button"
                onClick={props.backStep}
                className="margin-top-1"
            >
                {t("BackButton")}
            </Button>
            <Button type="button" onClick={onAdvanceStep} className="margin-top-1">
                {t("ContinueButton")}
            </Button>
            <Button
                variant="unstyled"
                className="text-base-dark hover:text-base-darker active:text-base-darkest margin-top-1"
                type="button"
                onClick={props.onCancel}
            >
                {t("Cancel")}
            </Button>
        </>
    );
}

export default ChooseData;
