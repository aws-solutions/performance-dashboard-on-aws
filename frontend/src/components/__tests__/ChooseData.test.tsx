/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import ChooseData from "../ChooseData";
import { DatasetType, SourceType } from "../../models";
import { MemoryRouter } from "react-router-dom";

test("renders the ChooseData component", async () => {
    const wrapper = render(
        <ChooseData
            selectDynamicDataset={() => {}}
            dynamicDatasets={[
                {
                    id: "123",
                    fileName: "abc",
                    s3Key: {
                        raw: "",
                        json: "abc.json",
                    },
                    sourceType: SourceType.IngestApi,
                    updatedAt: new Date("2020-01-01T00:00:00Z"),
                },
            ]}
            datasetType={DatasetType.DynamicDataset}
            onFileProcessed={() => {}}
            handleChange={() => {}}
            backStep={() => {}}
            advanceStep={() => {}}
            fileLoading={false}
            browseDatasets={() => {}}
            hasErrors={false}
            csvErrors={[]}
            csvFile={undefined}
            onCancel={() => {}}
            register={() => {}}
            widgetType="chart"
            dynamicFileName=""
            staticFileName=""
            setShowNoDatasetTypeAlert={() => {}}
        />,
        { wrapper: MemoryRouter },
    );
    expect(wrapper.container).toMatchSnapshot();
});
