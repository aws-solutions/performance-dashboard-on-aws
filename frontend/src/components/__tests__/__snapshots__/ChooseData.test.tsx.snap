// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the ChooseData component 1`] = `
<div>
  <fieldset
    class="usa-fieldset"
  >
    <legend
      class="usa-hint grid-col-6"
    >
      <label
        class="usa-label text-bold"
      >
        Data
      </label>
      <span
        class="usa-hint"
      >
        Choose an existing dataset or create a new one to populate this chart.
         
        <a
          class="usa-link action-link"
          href="/admin/apihelp"
          target="_blank"
        >
          <span>
            How do I add datasets?
          </span>
          <svg
            aria-hidden="false"
            aria-label="(opens in a new tab)"
            class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs margin-left-05"
            data-icon="external-link-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
              fill="currentColor"
            />
          </svg>
        </a>
      </span>
    </legend>
    <div
      class="grid-row"
    >
      <div
        class="padding-2px grid-row"
        role="radiogroup"
      >
        <div
          class="tablet:grid-col-4 padding-1"
        >
          <div
            class="usa-radio"
            role="radio"
          >
            <div
              class="tablet:grid-col"
            >
              <input
                aria-describedby="staticDataset-description"
                class="usa-radio__input usa-radio__input--tile radio"
                data-testid="staticDatasetRadioButton"
                id="staticDataset"
                name="datasetType"
                type="radio"
                value="StaticDataset"
              />
              <label
                class="usa-radio__label"
                for="staticDataset"
              >
                Static dataset
                <span
                  class="text-base usa-prose usa-checkbox__label-description"
                  id="staticDataset-description"
                >
                  Upload a new dataset from file or elect an existing dataset.
                </span>
              </label>
            </div>
          </div>
        </div>
        <div
          class="tablet:grid-col-4 padding-1"
        >
          <div
            class="usa-radio"
            role="radio"
          >
            <div
              class="tablet:grid-col"
            >
              <input
                aria-describedby="dynamicDataset-description"
                class="usa-radio__input usa-radio__input--tile radio"
                data-testid="dynamicDatasetRadioButton"
                id="dynamicDataset"
                name="datasetType"
                type="radio"
                value="DynamicDataset"
              />
              <label
                class="usa-radio__label"
                for="dynamicDataset"
              >
                Dynamic dataset
                <span
                  class="text-base usa-prose usa-checkbox__label-description"
                  id="dynamicDataset-description"
                >
                  Choose from a list of continuously updated datasets.
                </span>
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-label="Choose static dataset"
      hidden=""
      role="tabpanel"
      tabindex="0"
    >
      <div
        class="grid-row"
      >
        <div
          class="tablet:grid-col-6"
        >
          <div
            class="usa-form-group"
          >
            <label
              class="usa-label text-bold"
              for="dataset"
            >
              Static datasets
            </label>
            <div
              class="usa-hint"
            >
              <span>
                Upload a dataset from a CSV or Excel file, or choose an existing static dataset.
                 
                <a
                  class="usa-link action-link"
                  href="/admin/formatting"
                  target="_blank"
                >
                  <span>
                    How do I format my file?
                  </span>
                  <svg
                    aria-hidden="false"
                    aria-label="(opens in a new tab)"
                    class="svg-inline--fa fa-external-link-alt fa-w-16 fa-xs margin-left-05"
                    data-icon="external-link-alt"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                      fill="currentColor"
                    />
                  </svg>
                </a>
              </span>
            </div>
            <span
              class="usa-error-message"
              id="file-input-error-alert"
              role="alert"
            />
            <div
              class="usa-file-input"
            >
              <div
                class="usa-file-input__target"
              >
                <div
                  aria-hidden="true"
                  class="usa-file-input__instructions"
                >
                  <span
                    class="usa-file-input__drag-text"
                  >
                    Drag file here or 
                  </span>
                  <span
                    class="usa-file-input__choose"
                  >
                    choose from folder
                  </span>
                </div>
                <div
                  class="usa-file-input__box"
                />
                <input
                  accept=".csv,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
                  class="usa-file-input__input"
                  id="dataset"
                  name="dataset"
                  type="file"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="tablet:grid-col-3"
        >
          <button
            class="usa-button usa-button--outline vertCenter"
            type="button"
          >
            Browse datasets
          </button>
        </div>
      </div>
    </div>
    <div
      aria-label="Choose dynamic dataset"
      role="tabpanel"
      tabindex="0"
    >
      <div
        class="grid-row margin-top-3 margin-bottom-1"
      >
        <div
          class="tablet:grid-col-4 padding-top-1px"
        >
          <label
            for="search"
          >
            Search dynamic datasets
          </label>
          <div
            class="usa-search usa-search--small"
            role="search"
          >
            <input
              aria-label="Enter search terms"
              class="usa-input"
              id="search"
              name="query"
              placeholder=""
              style="height: 37px;"
              type="search"
              value=""
            />
            <button
              aria-label="Perform search"
              class="usa-button usa-button--base padding-x-2"
              style="height: 37px; border-top-left-radius: 0; border-bottom-left-radius: 0;"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-search fa-w-16 "
                data-icon="search"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="margin-top: -3px;"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="overflow-hidden"
      >
        <div
          class="overflow-x-hidden overflow-y-hidden"
        >
          <table
            class="usa-table usa-table--borderless"
            id="items"
            role="table"
            width="100%"
          >
            <thead>
              <tr
                role="row"
              >
                <th
                  aria-sort="none"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                  style="padding: 0.5rem 1rem;"
                >
                  <span />
                </th>
                <th
                  aria-sort="none"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                  style="padding: 0.5rem 1rem;"
                >
                  <button
                    aria-label="Toggle SortBy: Name"
                    class="usa-button usa-button--unstyled text-no-underline"
                    style="cursor: pointer;"
                    title="Toggle SortBy: Name"
                    type="button"
                  >
                    <span>
                      Name
                    </span>
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                      data-icon="chevron-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                        fill="currentColor"
                      />
                    </svg>
                  </button>
                </th>
                <th
                  aria-sort="descending"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                  style="padding: 0.5rem 1rem;"
                >
                  <button
                    aria-label="Toggle SortBy: Last updated"
                    class="usa-button usa-button--unstyled text-no-underline"
                    style="cursor: pointer;"
                    title="Toggle SortBy: Last updated"
                    type="button"
                  >
                    <span>
                      Last updated
                    </span>
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-chevron-circle-down fa-w-16 margin-left-1"
                      data-icon="chevron-circle-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zM273 369.9l135.5-135.5c9.4-9.4 9.4-24.6 0-33.9l-17-17c-9.4-9.4-24.6-9.4-33.9 0L256 285.1 154.4 183.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L239 369.9c9.4 9.4 24.6 9.4 34 0z"
                        fill="currentColor"
                      />
                    </svg>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                  style="padding: 0.5rem 1rem;"
                >
                  <button
                    aria-label="Toggle SortBy: Description"
                    class="usa-button usa-button--unstyled text-no-underline"
                    style="cursor: pointer;"
                    title="Toggle SortBy: Description"
                    type="button"
                  >
                    <span>
                      Description
                    </span>
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                      data-icon="chevron-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                        fill="currentColor"
                      />
                    </svg>
                  </button>
                </th>
                <th
                  aria-sort="none"
                  colspan="1"
                  role="columnheader"
                  scope="col"
                  style="padding: 0.5rem 1rem;"
                >
                  <button
                    aria-label="Toggle SortBy: Tags"
                    class="usa-button usa-button--unstyled text-no-underline"
                    style="cursor: pointer;"
                    title="Toggle SortBy: Tags"
                    type="button"
                  >
                    <span>
                      Tags
                    </span>
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                      data-icon="chevron-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                        fill="currentColor"
                      />
                    </svg>
                  </button>
                </th>
              </tr>
            </thead>
            <tbody
              role="rowgroup"
            >
              <tr
                role="row"
              >
                <td
                  role="cell"
                >
                  <div>
                    <input
                      style="cursor: pointer;"
                      type="radio"
                    />
                  </div>
                </td>
                <td
                  role="cell"
                >
                  abc
                </td>
                <td
                  role="cell"
                >
                  2020-01-01 00:00
                </td>
                <td
                  role="cell"
                >
                  
                </td>
                <td
                  role="cell"
                >
                  
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </fieldset>
  <br />
  <br />
  <hr />
  <button
    class="usa-button usa-button--outline margin-top-1"
    type="button"
  >
    Back
  </button>
  <button
    class="usa-button usa-button--base margin-top-1"
    type="button"
  >
    Continue
  </button>
  <button
    class="usa-button usa-button--unstyled text-base-dark hover:text-base-darker active:text-base-darkest margin-top-1"
    type="button"
  >
    Cancel
  </button>
</div>
`;
